### PR TITLE
Make "opam install --check" check if all dependencies are installed recursively

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Fixes https://github.com/ocaml/opam/issues/6097
Queued on top of #6121 
Queued on https://github.com/ocaml/opam/pull/6123

### CODE REVIEW

File: src/client/opamClient.ml [hunk 1]
```diff
@@ -1945,10 +1945,9 @@ let init
   gt, rt, default_compiler
 
 let check_installed ~build ~post t atoms =
-  let available = (Lazy.force t.available_packages) in
   let pkgs =
     OpamPackage.to_map
-      (OpamFormula.packages_of_atoms available atoms)
+      (OpamFormula.packages_of_atoms t.packages atoms)
```
Comment (rjbou) [inline src/client/opamClient.ml]: The performance improvement needs to be in its own PR.

Comment (rjbou) [reply src/client/opamClient.ml]: In fact, as the code is removed in the rest of the PR, shouldn't we not keep the commit?

**Review by rjbou — COMMENTED** [review]
lgtm on the idea ans its proper code.
I've added a test for basic cases for install check, like that we are sure that these changes don't change them. I don't know if it should be in this PR or #6121.

File: tests/reftests/install-check.test [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+The following actions will be faked:
+=== upgrade 1 package
+  - upgrade a 1 to 3
+=== install 1 package
+  - install b 1      [required by a]
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+Faking installation of b.1
+Faking installation of a.3
+Done.
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply tests/reftests/install-check.test:131]: no other reason other than the force of habit and just in case it makes it faster

File: src/client/opamClient.ml (hunk 1 above)
Comment (kit-ty-kate) [reply src/client/opamClient.ml]: the code is moved elsewhere in the final commit and is here to offset the performance cost of looking at the dependencies recursively

**Comment by kit-ty-kate** [comment]
The benchmark failure is real and is caused by an infinit loop. I'll have a look

File: master_changes.md [hunk 3]
```diff
@@ -212,6 +212,7 @@ users)
 
 ## opam-format
   * Add `OpamTypesBase.switch_selections_{compare,equal}`: proper comparison functions for `OpamTypes.switch_selections` [#6102 @kit-ty-kate]
+  * `OpamFormula`: add some missing comparison functions for `relop`, `version_constraint` and `atom` (`compare_relop`, `compare_version_constraint` and `compare_atom` respecively [#6122 @kit-ty-kate]
```
Comment (rjbou) [inline master_changes.md:216]:
```suggestion
  * `OpamFormula`: add some missing comparison functions for `relop`, `version_constraint` and `atom` (`compare_relop`, `compare_version_constraint` and `compare_atom` respectively) [#6122 @kit-ty-kate]

File: src/client/opamClient.ml (hunk 1 above)
Comment (rjbou) [reply src/client/opamClient.ml]: lgtm

**Review by rjbou — APPROVED** [review]
I've added some test from an old branch that I have, and some format nitpicking. Feel free to squash if it's ok for you.
Otherwise, lgtm!

File: src/client/opamClient.ml [hunk 4]
```diff
@@ -1962,62 +1957,109 @@ let check_installed ~build ~post t atoms =
...
+            then (new_map, seen)
+            else loop new_map seen found_conj
+      ) (map, seen) atoms
+  in
+  fst (loop OpamPackage.Name.Map.empty SeenSet.empty atoms)
 
 let assume_built_restrictions ?available_packages t atoms =
-  let missing = check_installed ~build:false ~post:false t atoms in
+  let missing =
+    check_installed ~build:false ~post:false ~recursive:false t atoms
```
Comment (rjbou) [inline src/client/opamClient.ml:2044]: Assume built does not check recursively, is it to not change the known behaviour?

Comment (kit-ty-kate) [reply src/client/opamClient.ml:2044]: no, the previous behaviour of this function did not check recursively so the behaviour of `assume-built` hasn't changed either

**Review by rjbou — APPROVED** [review]
Thanks!
